Load API Objects without contacting cluster

lilatomic opened the pull request:
This MR makes it possible to load the API objects into the `Api` without contacting the server.

Exposing a way to load it was easy. I also implemented loading from the kubectl cache as a sample client of that.

Implementing it, though, I felt it was a bit weird. Because the cache is on the whole list of API objects, it's not possible to really have a read-though cache. I'm wondering if it would be worth segmenting the cache? maybe the keys would be the name, kind, singularName, and shortNames, since those are used to search in `async_lookup_kind`. Thoughts?

fixes #722

lilatomic committed:
feature: expose cache in asyncache

lilatomic committed:
refactor: extract helper for formatting api resources

lilatomic committed:
feature: function to load api resources from kubectl cache

lilatomic committed:
refactor: glob for clearer and more accurate traversal

lilatomic committed:
refactor: add error handling code and debug logs

lilatomic commented:
Ah, so, the test failures are because the function the tests are using is a boundmethod of the class, and you can't set arbitrary attributes on those. You can set arbitrary attributes on functions (at the root or unbound on a class). That's how  it works for everything but the tests.

lilatomic committed:
wip: try reimplementing cache

We didn't actually use the features of asyncache (such as locking).
I could have figured out how to modify asyncache,
but it didn't support some of the usecases we'd want.
For example, it would be difficult to load it from kubectl and then reload it from the cluster if a value was missing.

lilatomic committed:
fixup! wip: try reimplementing cache

lilatomic committed:
fix: implement the rest of the kubectl cache finding logic

lilatomic committed:
refactor: pushdown cache-loaded logic to cache itself

lilatomic committed:
refactor: extract on-disk cache

lilatomic committed:
refactor: rename

lilatomic committed:
refactor: extrac calls to k8s to fetch kinds

will allow us to make a cached version

lilatomic committed:
feature: add KindFetcher for caching

lilatomic committed:
refactor: pushdown orchestration of kind fetching to ResourceKindCache

lilatomic committed:
refactor: collapse KindFetcherDisk and fold into KindFetcherCached

lilatomic committed:
refactor: collapse KindFetcher

lilatomic committed:
refactor: collapse KindFetcher hierarchy

lilatomic committed:
refactor: pushdown collecting kinds to Cache

will make bypassing cache easier

lilatomic committed:
refactor: extract helper function to handle cache in a consistent way

lilatomic committed:
fix: uncached fetch bypasses all cache

jacobtomlinson reviewed (commented):
Could you merge in from main, it shouldudl clean up the SPDX headers in the diff.

lilatomic committed:
fix: cap kubernetes-asyncio version for Python 3.9

lilatomic committed:
test: tests for caching

lilatomic committed:
fix: caching is never fatal

lilatomic committed:
fix: allow api requests without terminal `/`

kubernetes responded differently to some endpoints if they're `/apis` or `/apis/`. Making

lilatomic committed:
test: isolate kubectl cache dir

lilatomic committed:
feature: implement aggregated discovery

TODO: read from the cache

lilatomic committed:
refactor: cleanup

inline values, extract methods, real names

lilatomic committed:
feature: caching that checks the entire aggregate grain

lilatomic committed:
doc: update comments
